README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# {msm.stacked}

<!-- badges: start -->

[![R-CMD-check](https://github.com/RedDoorAnalytics/msm.stacked/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/RedDoorAnalytics/msm.stacked/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The {msm.stacked} package can be used to simplify the calculation of
state transition probabilities over time and the creation of stacked
probabilities plot from multi-state model fits from the
[{msm}](https://CRAN.R-project.org/package=msm) package.

# Installation

You can install the development version of {msm.stacked} from
[GitHub](https://github.com/RedDoorAnalytics/msm.stacked) with:

``` r
# install.packages("devtools")
devtools::install_github("RedDoorAnalytics/msm.stacked", build_vignettes = TRUE)
```

# Example

To illustrate the functionality of {msm.stacked}, we build upon the
documentation for the `msm::msm()` function. We will be using the heart
transplant data:

``` r
library(msm)
head(cav)
#>    PTNUM      age    years dage sex pdiag cumrej state firstobs statemax
#> 1 100002 52.49589 0.000000   21   0   IHD      0     1        1        1
#> 2 100002 53.49863 1.002740   21   0   IHD      2     1        0        1
#> 3 100002 54.49863 2.002740   21   0   IHD      2     2        0        2
#> 4 100002 55.58904 3.093151   21   0   IHD      2     2        0        2
#> 5 100002 56.49589 4.000000   21   0   IHD      3     2        0        2
#> 6 100002 57.49315 4.997260   21   0   IHD      3     3        0        3
```

Further details on this example dataset are included in the [vignette of
the {msm}
package](https://CRAN.R-project.org/package=msm/vignettes/msm-manual.pdf).

We start with a matrix of possible transitions:

``` r
twoway4.q <- rbind(
  c(-0.5, 0.25, 0, 0.25),
  c(0.166, -0.498, 0.166, 0.166),
  c(0, 0.25, -0.5, 0.25),
  c(0, 0, 0, 0)
)
twoway4.q
#>        [,1]   [,2]   [,3]  [,4]
#> [1,] -0.500  0.250  0.000 0.250
#> [2,]  0.166 -0.498  0.166 0.166
#> [3,]  0.000  0.250 -0.500 0.250
#> [4,]  0.000  0.000  0.000 0.000
```

This is then used to provide starting values for the model without
additional covariates:

``` r
cav.msm <- msm(
  formula = state ~ years,
  subject = PTNUM,
  data = cav,
  qmatrix = twoway4.q,
  deathexact = 4
)
cav.msm
#> 
#> Call:
#> msm(formula = state ~ years, subject = PTNUM, data = cav, qmatrix = twoway4.q,     deathexact = 4)
#> 
#> Maximum likelihood estimates
#> 
#> Transition intensities
#>                   Baseline                    
#> State 1 - State 1 -0.17037 (-0.19027,-0.15255)
#> State 1 - State 2  0.12787 ( 0.11135, 0.14684)
#> State 1 - State 4  0.04250 ( 0.03412, 0.05294)
#> State 2 - State 1  0.22512 ( 0.16755, 0.30247)
#> State 2 - State 2 -0.60794 (-0.70880,-0.52143)
#> State 2 - State 3  0.34261 ( 0.27317, 0.42970)
#> State 2 - State 4  0.04021 ( 0.01129, 0.14324)
#> State 3 - State 2  0.13062 ( 0.07952, 0.21457)
#> State 3 - State 3 -0.43710 (-0.55292,-0.34554)
#> State 3 - State 4  0.30648 ( 0.23822, 0.39429)
#> 
#> -2 * log-likelihood:  3968.798
```

We can use the `plot.msm()` function to plot survival curves from every
transient state to the final, absorbing state (e.g., a state denoting
death). This is denoted in the `cav` dataset by *state 4*:

``` r
plot(cav.msm, from = 1:3, to = 4)
```

<img src="man/figures/README-plot.msm-1.png" width="100%" />

The {msm} package also provides functionality to calculate state
transition probabilities at a given point in time. Say we are interested
in estimating the probability of being in a given state, from each
state, five years after baseline; we can use the `pmatrix.msm()`
function to obtain just that:

``` r
pmatrix.msm(x = cav.msm, t = 5)
#>            State 1    State 2    State 3   State 4
#> State 1 0.51965804 0.13851775 0.09119847 0.2506257
#> State 2 0.24386420 0.13881410 0.18090731 0.4364144
#> State 3 0.06121333 0.06897186 0.16909991 0.7007149
#> State 4 0.00000000 0.00000000 0.00000000 1.0000000
```

This shows that, for instance, study participants in State 1 at time
zero have (approximately) a 52% probability of still being in State 1
after years, 14% probability of being in State 2, 9% probability of
being in State 3, and 25% probability of being in State 4.

We can repeatedly call the `pmatrix.msm()` function to obtain
probabilities over time, but that’s a bit tedious. This is where the
{msm.stacked} package comes in handy.

Specifically, we can use the `stacked.data.msm()` function to calculate
transition probabilities over time, say, at 1 to 5 years:

``` r
library(msm.stacked)
sdd <- stacked.data.msm(model = cav.msm, tstart = 0, tforward = 5, tseqn = 6)
str(sdd)
#> 'data.frame':    96 obs. of  5 variables:
#>  $ from  : Factor w/ 4 levels "State 1","State 2",..: 1 2 3 4 1 2 3 4 1 2 ...
#>  $ to    : Factor w/ 4 levels "State 1","State 2",..: 1 1 1 1 2 2 2 2 3 3 ...
#>  $ p     : num  1 0 0 0 0 1 0 0 0 0 ...
#>  $ tstart: num  0 0 0 0 0 0 0 0 0 0 ...
#>  $ t     : num  0 0 0 0 0 0 0 0 0 0 ...
```

This returns a tidy dataset with all transition probabilities, from and
to every state, over `tseqn = 6` equally-spaced time intervals between
time zero and time five. Focussing on transitions from State 1 only:

``` r
subset(sdd, sdd$from == "State 1")
#>       from      to          p tstart t
#> 1  State 1 State 1 1.00000000      0 0
#> 5  State 1 State 2 0.00000000      0 0
#> 9  State 1 State 3 0.00000000      0 0
#> 13 State 1 State 4 0.00000000      0 0
#> 17 State 1 State 1 0.85395872      0 1
#> 21 State 1 State 2 0.08836953      0 1
#> 25 State 1 State 3 0.01475543      0 1
#> 29 State 1 State 4 0.04291632      0 1
#> 33 State 1 State 1 0.74313989      0 2
#> 37 State 1 State 2 0.12669585      0 2
#> 41 State 1 State 3 0.04053779      0 2
#> 45 State 1 State 4 0.08962646      0 2
#> 49 State 1 State 1 0.65472323      0 3
#> 53 State 1 State 2 0.14064466      0 3
#> 57 State 1 State 3 0.06380519      0 3
#> 61 State 1 State 4 0.14082692      0 3
#> 65 State 1 State 1 0.58161960      0 4
#> 69 State 1 State 2 0.14256253      0 4
#> 73 State 1 State 3 0.08072247      0 4
#> 77 State 1 State 4 0.19509541      0 4
#> 81 State 1 State 1 0.51965804      0 5
#> 85 State 1 State 2 0.13851775      0 5
#> 89 State 1 State 3 0.09119847      0 5
#> 93 State 1 State 4 0.25062574      0 5
```

Here we see, for instance, that the probability of still being in State
1, starting from State 1, is (approximately) 85% after one year, 74%
after two years, 65% after three years, 58% after four years, and 52%
after five years:

``` r
subset(sdd, sdd$from == "State 1" & sdd$to == "State 1")
#>       from      to         p tstart t
#> 1  State 1 State 1 1.0000000      0 0
#> 17 State 1 State 1 0.8539587      0 1
#> 33 State 1 State 1 0.7431399      0 2
#> 49 State 1 State 1 0.6547232      0 3
#> 65 State 1 State 1 0.5816196      0 4
#> 81 State 1 State 1 0.5196580      0 5
```

The package also provides functionality to automatically produce stacked
probabilities plots, for transition probabilities from and to every
state. This is implemented in the `stacked.plot.msm()` function:

``` r
stacked.plot.msm(model = cav.msm, tstart = 0, tforward = 5)
```

<img src="man/figures/README-stacked.plot-1.png" width="100%" />

This relies on [{ggplot2}](https://ggplot2.tidyverse.org) functionality
and returns a standard `ggplot` object, which can of course be further
customised beyond the default settings:

``` r
library(ggplot2)

stacked.plot.msm(model = cav.msm, tstart = 0, tforward = 5) +
  scale_fill_viridis_d(option = "plasma") +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(fill = "To:")
```

<img src="man/figures/README-stacked.plot.custom-1.png" width="100%" />

# Model with Covariates

We can of course incorporate covariates in a multi-state model and
obtain predictions for a specific covariates pattern; let’s demonstrate
this by incorporating sex in the model above. First, we fit a second
model:

``` r
cav.msm.cov <- msm(
  formula = state ~ years,
  subject = PTNUM,
  data = cav,
  covariates = ~sex,
  qmatrix = twoway4.q,
  deathexact = 4
)
cav.msm.cov
#> 
#> Call:
#> msm(formula = state ~ years, subject = PTNUM, data = cav, qmatrix = twoway4.q,     covariates = ~sex, deathexact = 4)
#> 
#> Maximum likelihood estimates
#> Baselines are with covariates set to their means
#> 
#> Transition intensities with hazard ratios for each covariate
#>                   Baseline                        
#> State 1 - State 1 -0.16938 (-1.894e-01,-1.515e-01)
#> State 1 - State 2  0.12745 ( 1.108e-01, 1.466e-01)
#> State 1 - State 4  0.04193 ( 3.354e-02, 5.241e-02)
#> State 2 - State 1  0.22645 ( 1.686e-01, 3.042e-01)
#> State 2 - State 2 -0.58403 (-1.053e+00,-3.238e-01)
#> State 2 - State 3  0.33693 ( 2.697e-01, 4.209e-01)
#> State 2 - State 4  0.02065 ( 2.196e-09, 1.941e+05)
#> State 3 - State 2  0.13050 ( 7.830e-02, 2.175e-01)
#> State 3 - State 3 -0.44178 (-5.582e-01,-3.497e-01)
#> State 3 - State 4  0.31128 ( 2.425e-01, 3.996e-01)
#>                   sex                            
#> State 1 - State 1                                
#> State 1 - State 2 0.5632779 (3.333e-01,9.518e-01)
#> State 1 - State 4 1.1289701 (6.262e-01,2.035e+00)
#> State 2 - State 1 1.2905854 (4.916e-01,3.388e+00)
#> State 2 - State 2                                
#> State 2 - State 3 1.0765518 (5.194e-01,2.231e+00)
#> State 2 - State 4 0.0003805 (7.241e-65,1.999e+57)
#> State 3 - State 2 1.0965531 (1.345e-01,8.937e+00)
#> State 3 - State 3                                
#> State 3 - State 4 2.4135380 (1.176e+00,4.952e+00)
#> 
#> -2 * log-likelihood:  3954.777
```

Then, we can use the same functionality as before to obtain stacked
probabilities plots:

``` r
stacked.plot.msm(model = cav.msm.cov, tstart = 0, tforward = 5) +
  labs(title = "Predictions for average covariates")
```

<img src="man/figures/README-stacked.plot.cov-1.png" width="100%" />

By default, this will set all covariates to their average value (as in
`pmatrix.msm()`); we can, however, pass specific covariates patterns
that we want to predict for:

``` r
stacked.plot.msm(model = cav.msm.cov, tstart = 0, tforward = 5, covariates = list(sex = 0)) +
  labs(title = "Predictions for 'sex = 0'")
```

<img src="man/figures/README-stacked.plot.cov.sex0-1.png" width="100%" />

``` r
stacked.plot.msm(model = cav.msm.cov, tstart = 0, tforward = 5, covariates = list(sex = 1)) +
  labs(title = "Predictions for 'sex = 1'")
```

<img src="man/figures/README-stacked.plot.cov.sex1-1.png" width="100%" />

This way we can provide clinically meaningful predictions that highlight
the effect of covariates of interest on state occupancy probabilities
over time.

# Model with Piecewise-Constant Intensities

By default, the {msm} package assumes constant (i.e., exponential)
baseline transition intensities. This means that predictions at t years
will be the same, irrespectively of when the starting point is:

``` r
stacked.plot.msm(model = cav.msm, tstart = 0, tforward = 3)
```

<img src="man/figures/README-stacked.plot.two.times-1.png" width="100%" />

``` r
stacked.plot.msm(model = cav.msm, tstart = 5, tforward = 3, start0 = FALSE)
```

<img src="man/figures/README-stacked.plot.two.times-2.png" width="100%" />

We can relax this assumption by allowing piecewise-constant baseline
transition rates. This can be done by setting the `pci` argument of
`msm()`:

``` r
cav.msm.pw <- msm(
  formula = state ~ years,
  subject = PTNUM,
  data = cav,
  qmatrix = twoway4.q,
  deathexact = 4,
  pci = quantile(x = cav$years, probs = c(0.25, 0.50, 0.75))
)
#> Warning in msm(formula = state ~ years, subject = PTNUM, data = cav, qmatrix =
#> twoway4.q, : Optimisation has probably not converged to the maximum likelihood
#> - Hessian is not positive definite.
cav.msm.pw
#> 
#> Call:
#> msm(formula = state ~ years, subject = PTNUM, data = cav, qmatrix = twoway4.q,     deathexact = 4, pci = quantile(x = cav$years, probs = c(0.25,         0.5, 0.75)))
#> 
#> Optimisation probably not converged to the maximum likelihood.
#> optim() reported convergence but estimated Hessian not positive-definite.
#> 
#> -2 * log-likelihood:  3887.911
```

Specifically, here we set cut-points at quartiles of the observed
distribution of (possibly censored) transition times. We ignore the
warning about *non-convergence* for now - in practice, we should
investigate this and try a different optimiser or “consider tightening
the tolerance criteria for convergence” (according to the documentation
of `msm()`).

First, we can do a likelihood ratio test to check whether the model with
piecewise-constant intensities fits the data better:

``` r
lrtest.msm(cav.msm, cav.msm.pw)
#>            -2 log LR df            p
#> cav.msm.pw   80.8872 21 5.736045e-09
```

The test is statistically significant at any usual level, thus the more
flexible model seems appropriate. Predictions of transition
probabilities will now depend on the starting point `tstart`, even
though `tforward` is the same:

``` r
stacked.plot.msm(model = cav.msm.pw, tstart = 0, tforward = 3)
```

<img src="man/figures/README-stacked.plot.two.times.pw-1.png" width="100%" />

``` r
stacked.plot.msm(model = cav.msm.pw, tstart = 5, tforward = 3, start0 = FALSE)
```

<img src="man/figures/README-stacked.plot.two.times.pw-2.png" width="100%" />

As expected, we see that the predicted probabilities between 0 and 3
units of time are now different compared to those between 5 and 8, given
the (now) non-constant baseline intensities.

# Excluding States

The {msm.stacked} package also include functionality to calculate (and
plot) transition probabilities from only certain states of interest. For
instance, the models we fit in the previous examples includes an
absorbing state, *State 4*, from which there will be no transitions. For
this example, we will be using the model with constant baseline
transition rates (`cav.msm`).

Let’s start with a utility function to determine the names of the state
of a {msm} model fit, which is included in {msm.stacked}. This is called
`states.msm()`:

``` r
states.msm(cav.msm)
#> [1] "State 1" "State 2" "State 3" "State 4"
```

Now we have the correct names used by {msm} to define each state. Let’s
calculate transitions probabilities from all states, but excluding the
absorbing *State 4*:

``` r
stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 3, exclude = "State 4")
#>       from      to           p tstart   t
#> 1  State 1 State 1 1.000000000      0 0.0
#> 2  State 2 State 1 0.000000000      0 0.0
#> 3  State 3 State 1 0.000000000      0 0.0
#> 4  State 1 State 2 0.000000000      0 0.0
#> 5  State 2 State 2 1.000000000      0 0.0
#> 6  State 3 State 2 0.000000000      0 0.0
#> 7  State 1 State 3 0.000000000      0 0.0
#> 8  State 2 State 3 0.000000000      0 0.0
#> 9  State 3 State 3 1.000000000      0 0.0
#> 10 State 1 State 4 0.000000000      0 0.0
#> 11 State 2 State 4 0.000000000      0 0.0
#> 12 State 3 State 4 0.000000000      0 0.0
#> 13 State 1 State 1 0.921422669      0 0.5
#> 14 State 2 State 1 0.093120696      0 0.5
#> 15 State 3 State 1 0.003009286      0 0.5
#> 16 State 1 State 2 0.052893658      0 0.5
#> 17 State 2 State 2 0.745001419      0 0.5
#> 18 State 3 State 2 0.050466555      0 0.5
#> 19 State 1 State 3 0.004483375      0 0.5
#> 20 State 2 State 3 0.132369477      0 0.5
#> 21 State 3 State 3 0.808061596      0 0.5
#> 22 State 1 State 4 0.021200298      0 0.5
#> 23 State 2 State 4 0.029508408      0 0.5
#> 24 State 3 State 4 0.138462563      0 0.5
#> 25 State 1 State 1 0.853958721      0 1.0
#> 26 State 2 State 1 0.155576908      0 1.0
#> 27 State 3 State 1 0.009903994      0 1.0
#> 28 State 1 State 2 0.088369526      0 1.0
#> 29 State 2 State 2 0.566632840      0 1.0
#> 30 State 3 State 2 0.078536913      0 1.0
#> 31 State 1 State 3 0.014755432      0 1.0
#> 32 State 2 State 3 0.205995634      0 1.0
#> 33 State 3 State 3 0.659657266      0 1.0
#> 34 State 1 State 4 0.042916321      0 1.0
#> 35 State 2 State 4 0.071794618      0 1.0
#> 36 State 3 State 4 0.251901827      0 1.0
```

As you can see, transitions from the state passed to `exclude` are not
reported. We can also exclude more than one state, for instance if we
want to calculate only transitions from *State 1*:

``` r
stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 3, exclude = c("State 2", "State 3", "State 4"))
#>       from      to           p tstart   t
#> 1  State 1 State 1 1.000000000      0 0.0
#> 2  State 1 State 2 0.000000000      0 0.0
#> 3  State 1 State 3 0.000000000      0 0.0
#> 4  State 1 State 4 0.000000000      0 0.0
#> 5  State 1 State 1 0.921422669      0 0.5
#> 6  State 1 State 2 0.052893658      0 0.5
#> 7  State 1 State 3 0.004483375      0 0.5
#> 8  State 1 State 4 0.021200298      0 0.5
#> 9  State 1 State 1 0.853958721      0 1.0
#> 10 State 1 State 2 0.088369526      0 1.0
#> 11 State 1 State 3 0.014755432      0 1.0
#> 12 State 1 State 4 0.042916321      0 1.0
```

Of course, this functionality is also included in the plotting function:

``` r
stacked.plot.msm(model = cav.msm, tstart = 0, tforward = 10, exclude = "State 4") +
  theme(legend.position = "bottom")
```

<img src="man/figures/README-plot.no.absorbing-1.png" width="100%" />


NEWS.md
# {msm.stacked} 0.0.3

### Breaking changes

- Updated algorithm to estimate confidence intervals for transition probabilities to be more consistent across time-points.
  This is based on a parametric bootstrap approach, where we use each resample along the entire time requested for prediction, with confidence intervals then calculated using the percentile method.
  The arguments of `stacked.data.msm()` have been updated accordingly, and names have been adjusted to not collide with {msm}.
  Please read the documentation page at `?stacked.data.msm()` to get familiar with the new arguments.

### Housekeeping

- Removed a URL in the vignette on confidence intervals that was giving a `403 Forbidden` error. 

# {msm.stacked} 0.0.2

### New features

- It is now possible to calculate confidence intervals for the transition probabilities returned by `stacked.data.msm()` via the `ci` argument.
  This supports all methods implemented in `msm::pmatrix.msm()`, and can be used with multi-state models with and without covariates.
- Added new vignettes describing the functionality of {msm.stacked}.
  They can be found by typing the following in your R console:
  ```r
  vignette("A-introduction", package = "msm.stacked")
  vignette("B-ci", package = "msm.stacked")
  ```

### Housekeeping

- Updated README file.
- Updated copyright year.

# {msm.stacked} 0.0.1

Initial release of the package. Currently, the following functions are included:
- `stacked.data.msm()`, to calculate transition probabilities over time from an {msm} model fit;
- `stacked.plot.msm()`, to automatically produce stacked probabilities plots based on transition probabilities over time;
- `states.msm()`, a utility function to determine the names of states for an {msm} model.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(msm.stacked)

test_check("msm.stacked")


tests/testthat/test-stacked.data.msm.R
# Based on example in ?msm
library(msm)
twoway4.q <- rbind(
  c(-0.5, 0.25, 0, 0.25),
  c(0.166, -0.498, 0.166, 0.166),
  c(0, 0.25, -0.5, 0.25),
  c(0, 0, 0, 0)
)
cav.msm <- msm(
  formula = state ~ years,
  subject = PTNUM,
  data = cav,
  qmatrix = twoway4.q,
  deathexact = 4,
  fixedpars = TRUE # only to speed up examples!
)

test_that("Error checks", {
  #
  expect_error(object = stacked.data.msm())
  expect_error(object = stacked.data.msm(model = cav.msm))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0))
  expect_error(object = stacked.data.msm(model = cav.msm, tforward = 1))

  #
  expect_error(object = stacked.data.msm(model = cav, tstart = 0, tforward = 1))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = c(0, 1), tforward = 1))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = c(0, 1)))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 10:20))

  #
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = -1, tforward = 1))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = 0))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = -1))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 0))
  expect_error(object = stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 1))
})

test_that("Output", {
  sda <- stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 10)

  #
  expect_s3_class(object = sda, class = "data.frame")
  expect_true(object = inherits(x = sda, what = "data.frame"))

  #
  expect_equal(object = nrow(sda), expected = 10 * (ncol(twoway4.q)^2))
})


R/stacked.data.msm.R
#' @title Calculate State Occupancy Probabilities Over Time from and to Every State
#'
#' @description The [stacked.data.msm()] function can be used to calculate
#'   state occupancy probabilities over time for multi-state models fitted
#'   using the {msm} package.
#'
#' @param model A multi-state model fitted using [msm::msm()].
#' @param tstart Numeric value denoting the starting time for calculating predictions.
#'   It can be time zero (e.g., the beginning of the study) or it can be a given follow-up
#'   time (e.g., one year after baseline) in which case predictions will be conditional on
#'   being in a given state at time `tstart`.
#' @param tforward Numeric value denoting how many units of time forward predictions are
#'   to be calculated for. For instance, if `tstart = t0` and `tforward = t1`, predictions
#'   will be computed up to time `t0 + t1`.
#' @param tseqn Numeric value denoting how many sub-intervals to use between `tstart` and
#'   `tstart + tforward`, defaulting to 5. This must be greater than one.
#' @param exclude Denotes a state (or more than one) for which transitions from such state are
#'   to be excluded. This is useful, for example, to avoid returning transitions from an absorbing
#'   state; the [states.msm()] function can be helpful to identify the names of the states for a
#'   given model. The `exclude` parameter defaults to `NULL`, where all transitions are returned.
#' @param conf.int Boolean value denoting whether confidence intervals for transition probabilities
#'   over time should be returned. This uses a parametric bootstrap approach and the percentile method,
#'   similar to that of `ci = "normal"` in [msm::pmatrix.msm()], but modified to ensure that a single
#'   model resample is used throughout all time points.
#'   Defaults to `FALSE`.
#' @param B Number of replication for the parametric boostrap procedure used to calculate
#'   confidence intervals if `conf.int = TRUE`.
#'   Defaults to 1,000.
#' @param alpha Desired significance level (1 - confidence level) for the confidence intervals if
#'   `conf.int = TRUE`.
#'   Defaults to 0.05
#' @param progress Boolean denoting whether a progress bar should be printed to show the status of
#'   the parametric bootstrap.
#'   Defaults to `TRUE` if `conf.int = TRUE`, `FALSE` otherwise.
#' @param ... Additional arguments to be passed to [msm::pmatrix.msm()]. This is useful,
#'   for example, if calculating predictions for a certain covariates pattern - otherwise,
#'   as in [msm::pmatrix.msm()], predictions will be assuming means of the covariates in
#'   the data.
#'
#' @return
#' A data frame with the following columns:
#' - `from`, denoting the starting state;
#' - `to`, denoting the destination state;
#' - `p`, denoting the probability of being in state `to` starting from state `from` at
#'   time `t`;
#' - `tstart`, denoting the starting point for the predictions;
#' - `t`, denoting the times for the predicted probabilities `p`.
#'
#' If `conf.int = TRUE`, then two columns with the lower and upper confidence interval
#' bounds will be included as well: `conf.low`, `conf.high`.
#'
#' @seealso [msm::msm()], [msm::pmatrix.msm()]
#'
#' @export
#'
#' @examples
#'
#' ### Example 1:
#' # Based on example in ?msm
#' # We first fit a {msm} model without covariates
#' library(msm)
#' twoway4.q <- rbind(
#'   c(-0.5, 0.25, 0, 0.25),
#'   c(0.166, -0.498, 0.166, 0.166),
#'   c(0, 0.25, -0.5, 0.25),
#'   c(0, 0, 0, 0)
#' )
#' cav.msm <- msm(
#'   formula = state ~ years,
#'   subject = PTNUM,
#'   data = cav,
#'   qmatrix = twoway4.q,
#'   deathexact = 4
#' )
#'
#' # Predictions from time 0 to time 1, with 3 mid-points:
#' p1 <- stacked.data.msm(model = cav.msm, tstart = 0, tforward = 1, tseqn = 3)
#' head(p1)
#'
#' # Predictions from time 1 to time 5, with 5 mid-points:
#' p2 <- stacked.data.msm(model = cav.msm, tstart = 1, tforward = 4, tseqn = 5)
#' head(p2)
#'
#' ### Example 2:
#' # Model with covariates
#' cav.msm.cov <- msm(
#'   formula = state ~ years,
#'   subject = PTNUM,
#'   data = cav,
#'   covariates = ~sex,
#'   qmatrix = twoway4.q,
#'   deathexact = 4,
#'   fixedpars = TRUE # only to speed up examples!
#' )
#'
#' # Predictions from time 0 to time 5, mean covariates values:
#' p3 <- stacked.data.msm(model = cav.msm.cov, tstart = 0, tforward = 5)
#'
#' # Predictions from time 0 to time 5, for `sex = 0`:
#' p4 <- stacked.data.msm(model = cav.msm.cov, tstart = 0, tforward = 5, covariates = list(sex = 0))
#'
#' # Predictions from time 0 to time 5, for `sex = 1`:
#' p5 <- stacked.data.msm(model = cav.msm.cov, tstart = 0, tforward = 5, covariates = list(sex = 1))
#'
#' # p3, p4, and p5 should all be different:
#' all.equal(p3, p4)
#' all.equal(p3, p5)
#' all.equal(p4, p5)
#'
#' ### Example 3:
#' # Excluding transitions from a certain state, e.g., from State 4:
#' stacked.data.msm(model = cav.msm.cov, tstart = 0, tforward = 5, exclude = "State 4")
#'
#' # Returning transitions from State 1 only:
#' stacked.data.msm(
#'   model = cav.msm.cov,
#'   tstart = 0,
#'   tforward = 5,
#'   exclude = c("State 2", "State 3", "State 4")
#' )
#'
#' ### Example 4:
#' # Confidence intervals for transition probabilities,
#' # using B = 10 replicates to keep it fast:
#' p1.ci <- stacked.data.msm(
#'   model = cav.msm,
#'   tstart = 0,
#'   tforward = 1,
#'   tseqn = 3,
#'   conf.int = TRUE,
#'   B = 10,
#'   progress = FALSE
#' )
#' head(p1.ci)
#' # Compare with:
#' head(p1)
stacked.data.msm <- function(model, tstart, tforward, tseqn = 5, exclude = NULL, conf.int = FALSE, B = 1000, alpha = 0.05, progress = conf.int, ...) {
  # Check arguments
  arg_checks <- checkmate::makeAssertCollection()
  # 'model' must be of class 'msm'
  checkmate::assert_class(x = model, classes = "msm", add = arg_checks, .var.name = "model")
  # 'tstart', 'tforward', 'tseq', 'ref' must be a single number
  checkmate::assert_number(x = tstart, add = arg_checks, .var.name = "tstart")
  checkmate::assert_number(x = tforward, add = arg_checks, .var.name = "tforward")
  checkmate::assert_number(x = tseqn, add = arg_checks, .var.name = "tseqn")
  # 'exclude' must be a string (or vector of strings), but can be NULL
  checkmate::assert_character(x = exclude, add = arg_checks, null.ok = TRUE, .var.name = "exclude")
  # 'tseqn' must be greater than one
  checkmate::assert_true(x = (tseqn > 1), add = arg_checks, .var.name = "tseqn > 1")
  # 'tstart', 'tforward' must be greater than zero
  checkmate::assert_true(x = (tstart >= 0), add = arg_checks, .var.name = "tstart >= 0")
  checkmate::assert_true(x = (tforward > 0), add = arg_checks, .var.name = "tforward > 0")
  # 'ci' must be a single boolean
  checkmate::assert_logical(x = conf.int, len = 1, add = arg_checks, .var.name = "ci")
  # 'B' must be a single integer, greater than zero
  checkmate::assert_number(x = B, add = arg_checks, .var.name = "B")
  checkmate::assert_true(x = (B > 0), add = arg_checks, .var.name = "B > 0")
  # 'alpha' must be a single number between 0 and 1
  checkmate::assert_number(x = alpha, add = arg_checks, .var.name = "alpha")
  checkmate::assert_true(x = (alpha > 0 & alpha < 1), add = arg_checks, .var.name = "alpha in [0, 1]")
  # 'progress' must be a single boolean
  checkmate::assert_logical(x = progress, len = 1, add = arg_checks, .var.name = "progress")

  # Report
  if (!arg_checks$isEmpty()) checkmate::reportAssertions(arg_checks)

  # Sequence of `tseqn` equally-spaced points for forward predictions
  tseq <- seq(0, tforward, length.out = tseqn)

  # If conf.int = TRUE but model did not have asymptotic SEs we cannot do conf.int
  # So, we fail gracefully
  if (conf.int & !model$foundse) {
    message("Asymptotic standard errors not available in fitted model. Continuing with conf.int = FALSE...")
    conf.int <- FALSE
  }

  # If confidence intervals are required, we need to resample models once per value of tseq
  if (conf.int) {
    resamp_models <- .resample_models(x = model, B = B)
  }

  # Setup progress bar if required
  if (conf.int & progress) {
    pb <- utils::txtProgressBar(min = 0, max = B * length(tseq), style = 3)
  }

  # Calculate pmatrix at each time point forward
  preds <- lapply(X = tseq, FUN = function(.t) {
    # Point estimates
    point_estimate <- .wrap_pmatrix.msm(t = .t, model = model, t1 = tstart, ...)
    if (conf.int) {
      # Confidence intervals
      # Setup progress bar, if required
      repl_point_estimate <- sapply(X = resamp_models, FUN = function(x) {
        # Calculate values for this model replicate
        out <- .wrap_pmatrix.msm(t = .t, model = x, t1 = tstart, ...)
        # Increment progress bar
        if (progress) utils::setTxtProgressBar(pb = pb, value = pb$getVal() + 1)
        # Return
        return(out$p)
      }, simplify = "matrix")
      # Calculate confidence intervals using the quantile method
      ci <- matrixStats::rowQuantiles(x = repl_point_estimate, probs = c(alpha / 2, 1 - alpha / 2))
      colnames(ci) <- c("conf.low", "conf.high")
      # Combine
      out <- cbind(point_estimate, ci)
      # Return
      return(out)
    }
    # Else, return point estimates only
    return(point_estimate)
  })

  # Close progress bar
  if (conf.int & progress) close(pb)

  # Bind rows
  preds <- do.call(rbind.data.frame, preds)

  # Exclude states if 'exclude' was defined
  if (!is.null(exclude)) {
    preds <- subset(preds, !(preds$from %in% exclude))
  }

  # Remove row names
  rownames(preds) <- NULL

  # Return data
  return(preds)
}
